### PR TITLE
[WIP] Exact float parsing

### PR DESCRIPTION
**Pull request by lemire**
This is an attempt at providing exact (or more exact) float parsing at high speed.

- [ ] in functions like compute_float_64 and the 128-bit counterpart, we need to ensure that the binary exponent is always greater than 0 and strictly smaller than 0x7FF.
- [ ] we need to add more testing
- [ ] we should compare the performance against abseil-cpp, ScanadvDouble, Boost Spirit and Andrei Alexandrescu's implementation (Folly)
- [ ] the lookup tables could be made smaller, we should investigate
- [ ] Make sure that the code compiles under most compilers including Visual Studio; this involves avoiding 128-bit integers.
- [ ] Benchmark on ARM processors.

Fixes https://github.com/lemire/simdjson/issues/242

This replaces https://github.com/lemire/simdjson/pull/303

### CODE REVIEW

**Comment by lemire** [comment]
The current code cannot parse something like 1e-308 so this is a work in progress.

In compute_float_64, we compute real_exponent as c.exp + 1023 + (127 - lz). Presumably, lz is in [0,64] (it starts out in [0,64) but then gets incremented) but c.exp is from  power_of_ten_components and can go down to -1087. This means that we can get a negative value (-1087 + 1023 + 127 - 64 == -1). That's  almost surely a source of concern.

What is the powersOf10 array for?

Clearly the exponent in power_of_ten_components, we store the mantissa and the exponent separately, while it seems obvious that a 64-bit value would be enough. At a minimum,  the exponent  could be stored as a 16-bit value as opposed to a 32-bit value. In any case, we could save a little bit of space. Whether this is worthwhile should be tested out experimentally. 


In power_of_ten_components, why are the exponents stored as values ranging from -1087 to 960? They are later corrected with extra arithmetic operations. E.g. we add + 1023 + 127 to it. This seems clearly wasteful.

**Comment by lemire** [comment]
Ping @michaeleisel

**Comment by lemire** [comment]
@michaeleisel  Please have a look at my comments. I think we need to discuss this further.

**Comment by michaeleisel** [comment]
The current code cannot parse something like 1e-308 so this is a work in progress.

> In compute_float_64, we compute real_exponent as c.exp + 1023 + (127 - lz). Presumably, lz is in [0,64] (it starts out in [0,64) but then gets incremented) but c.exp is from power_of_ten_components and can go down to -1087. This means that we can get a negative value (-1087 + 1023 + 127 - 64 == -1). That's almost surely a source of concern.

Yeah that wouldn't surprise me. This code hasn't been tested with values that are not extremely small or large, it will need a bit of work there to cover edge cases like this

> What is the powersOf10 array for?

powersOf10 array? I don't see an array called `powersOf10`

> Clearly the exponent in power_of_ten_components, we store the mantissa and the exponent separately, while it seems obvious that a 64-bit value would be enough. At a minimum, the exponent could be stored as a 16-bit value as opposed to a 32-bit value. In any case, we could save a little bit of space. Whether this is worthwhile should be tested out experimentally.

You're saying it seems obvious a 64-bit value would be enough to store *both* mantissa and exponent? I don't think so, having 64 bits for the mantissa alone is important. But yes, the exponent could definitely be stored in a smaller data type.

> In power_of_ten_components, why are the exponents stored as values ranging from -1087 to 960? They are later corrected with extra arithmetic operations. E.g. we add + 1023 + 127 to it. This seems clearly wasteful.

Yes, this is wasteful, it can be fixed

cc @michaeleisel

**Comment by michaeleisel** [comment]
s/not extremely/extremely

**Comment by lemire** [comment]
> I don't see an array called powersOf10

Now you don't, see this commit : https://github.com/lemire/simdjson/pull/333/commits/71480bee6509b46fa2c25d62dfe265f62031438b

**Comment by lemire** [comment]
> having 64 bits for the mantissa alone is important

Ok.

**Comment by lemire** [comment]
> This code hasn't been tested with values that are not extremely small or large, it will need a bit of work there to cover edge cases like this

Ok.

**Comment by michaeleisel** [comment]
powersOfTen is a remnant of other code, it can be removed

**Comment by michaeleisel** [comment]
The testing will need some thinking through, and maybe a flag to force it to use the 128-bit function, as that’s an especially tricky function. Also, are you ok with the performance of this version, or is more optimization work planned?

**Comment by lemire** [comment]
@michaeleisel You have convinced me, with this code, that we can get much more accurate float parsing (maybe exact) with, in practice, only a modest (though not trivial) overhead. That is, it is fast enough that it could replace our default. It would be a performance regression, but we would buy something valuable in return without paying too much for it.

The worst case in our standard files is canada.json... We go from 1.1 GB/s to 0.9 GB/s. But this file is literally a number parsing benchmark by itself. It is likely a "practical" worst case (that is, you can making things much much worse, but that would probably require designing an adversarial example).

Assuming I am not being mislead, I will put it this way: "you proved me wrong". 

I am also cc ing  fellow hackers @jkeiser, @geofflangdale and others.

**Comment by lemire** [comment]
@michaeleisel I would say that, at this point, the priority is to make this code correct. A starting point is better testing.

Later we will need to "reason it out" and "prove it correct".

And then we can optimize the hell out of it.

**Comment by michaeleisel** [comment]
makes sense. btw we will have to rewrite the __uint_128 stuff for msvc. perhaps using a struct with two uint64_t's and some custom operators would make sense

**Comment by lemire** [comment]
@michaeleisel We do not actually need the __uint_128 type per se, and I know exactly how to write the code for Visual Studio, so it is not a concern for me. We can handle it later.

**Comment by lemire** [comment]
To be compared with 

1. [abseil-cpp](https://github.com/abseil/abseil-cpp/blob/master/absl/strings/charconv.h)
2. [ScanadvDouble](https://github.com/chrchang/plink-ng/blob/master/2.0/plink2_string.cc)
3. Boost Spirit
4. Andrei Alexandrescu's implementation (Folly)

And others.

**Comment by michaeleisel** [comment]
I’ll take a look at those soon. On the testing front, assuming it’s fast, https://blog.trailofbits.com/2019/11/27/64-bits-ought-to-be-enough-for-anybody/ has me thinking about testing all possible cases

**Comment by michaeleisel** [comment]
i haven't built/run anything, but here are my high-level thoughts:
- abseil appears to just use `compute_float_64` and not the other ones, and it looks like with extra work to do it
- ScanadvDouble appears to be inexact
- boost and folly look like template madness that will take more time to look at

**Comment by lemire** [comment]
@michaeleisel Thank you. I am planning to build an experimental assessment.

I do expect the Boost version to be slightly insane, such is Boost... and I really don't want to rely on Boost... but we might be able to learn from them if they have good algorithms.

**Comment by michaeleisel** [comment]
I'd be happy to help with this, not sure what your exact plan is

**Comment by lemire** [comment]
@michaeleisel Please do help!!!

There is a long to-do list that I drafted. 

I'd be happy to discuss some more.

If nothing else, further comparisons between existing solutions we could borrow would be great.

**Comment by lemire** [comment]
And, of course, the PR has gone out of sync with the rest of the code.

**Comment by lemire** [comment]
(Basically, if I am the only one working on this, it won't get done in a few days.)

**Comment by lemire** [comment]
@michaeleisel  

For now, see my comments. The compute_float_64, to begin with, is provably incorrect. I pointed out my concerns earlier, but a cursory code inspection shows that real_exponent can overflow its 11 bits. It is even trivial to come up with an example, just try parsing 1e-308.

So we need to fix these functions, and be confident that they are correct. So we need to go through and justify all our steps.

**Comment by lemire** [comment]
Note that the new code fails our tests, as CI shows. So this needs to be fixed, of course.

**Comment by lemire** [comment]
Ok. Now we at least pass *some* of the tests and the code is slightly less ugly.

**Comment by michaeleisel** [comment]
Yes, this is all proof-of-concept code :)

Thanks for this work to clean it up, I'll take a look soon

**Comment by lemire** [comment]
> Yes, this is all proof-of-concept code :)

I realize this, and I am not criticizing you. However, I am trying to document as clearly as possible where we are.

**Comment by michaeleisel** [comment]
Here's a performance comparison:

simdjson
168483
abseil
402654
boost spirit
278618
folly
1143387
folly 2
1113136
double-conversion
1045300

Looks like simdjson is the fastest, and Boost Spirit is the second fastest. ScanadvDouble was excluded because it's not precise (based on reading the code and and the comments). Google's double-conversion library is included, which is used by folly and taken from V8. Benchmarking code is at https://github.com/michaeleisel/strtodComp/ . Everything besides abseil was actually really easy to integrate (and I've included a static lib in the repo for abseil)

Now, I don't know if simdjson has a faster algorithm to get the mantissa (basically all the code before the part that I've changed). It's possible that it could compensate for slower performance in the part that I've changed, although probably not to verify either way.

**Comment by michaeleisel** [comment]
Another interesting library is https://github.com/Alexhuszagh/rust-lexical , but I figure it wouldn't be integratable anyways

**Comment by jkeiser** [comment]
Are the "simdjson" numbers there from this PR, or from master?

I am glad we're faster than so many libraries, but also kinda surprised. We're not really using tricks in there, besides maybe the 8-byte-at-a-time number parse check, which should generally only matter to ... benchmarks ... oh :)

**Comment by lemire** [comment]
@jkeiser

I think that @michaeleisel points to his repo and going there we find that he appears to be benchmarking this PR.

I am not surprised by the good performance. It is a neat piece of engineering based on layers of good heuristics.

It is also incorrect and untested. :-)

I think that Michael's point is that fixing it up and testing it up is worth the effort.

**Comment by michaeleisel** [comment]
Are we feeling confident enough about the performance to go forwards, or are there other implementations that would be worth comparing?

**Comment by lemire** [comment]
I think that the next step is to take each one of the parsers (since I view this number parser as a sequence of parsers) and prove them correct and test them carefully.

**Comment by michaeleisel** [comment]
I see - I feel like I've shown mathematically that they're sound, but is there a different sort of proof you're interested in?

**Comment by lemire** [comment]
@michaeleisel Given that we know that the implementation is not currently correct (among other things, it can overflow), then we know that it has not been proven correct. Maybe you mean that you have proven correct the conceptual idea? Yes, you did. But it is the implementation that we need to validate.

**Comment by lemire** [comment]
I plan to resume work on this within a week.

**Comment by lemire** [comment]
>  I've included a static lib in the repo for abseil

That's not likely to work (this is system specific).

**Comment by michaeleisel** [comment]
The static lib is just to benchmark abseil's parsing against ours

**Comment by lemire** [comment]
> The static lib is just to benchmark abseil's parsing against ours

I understand, but there is no reason to believe that such a precompiled library will work over different systems. It is likely specific to a version of macOS in your case.

I will propose a more portable approach.

**Comment by michaeleisel** [comment]
sounds good. you can do something roughly along the lines of:
```
	curl -# -L https://github.com/abseil/abseil-cpp/archive/20200225.tar.gz | tar xz
	mkdir $@/build
	cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_STANDARD=17 -S $@ -B $@/build
	make -C $@/build -j
	find $@/build/absl -name '*.a' | xargs libtool -static -o $@/build/libabsl.a
```

**Comment by michaeleisel** [comment]
it would be interesting to investigate how much of the speedup is just due to the faster code that precedes my changes, and if e.g. boost spirit has comparable performance to the changes that i've made and just not the stuff preceding

**Comment by lemire** [comment]
> it would be interesting to investigate how much of the speedup is just due to the faster code that precedes my changes, and if e.g. boost spirit has comparable performance to the changes that i've made and just not the stuff preceding

I am not following this.

**Comment by michaeleisel** [comment]
Like, there are two big components here: one is reading the double in as an integer, i.e. ignoring the decimal point, and the other is using that integer with various strategies to get a float from it. the changes i've made are specifically to the latter part. but the first part is important too, in that it may be faster in simdjson than it is in other libraries, and thus responsible for the fast performance we're seeing in benchmarks

**Comment by lemire** [comment]
@michaeleisel  Good point. It is hard to assess, but we can try to say something about it.

I am working on making the installation of the dependencies cross-platform and, honestly, abseil is the good citizen here. Everything else is a damn mess of manual labor.

**Comment by michaeleisel** [comment]
that's interesting, they were all super easy for me. i think i used homebrew for a bunch of them? maybe there's no easy equivalent for linux

**Comment by michaeleisel** [comment]
i can say a few that i pretty much know are *not* candidates for being fast though: anything which uses DiyFp, a strategy that i've seen in rapidjson: https://rapidjson.org/diyfp_8h_source.html (assuming that their DiyFp is the same as rapidjson's) . it's really slow. so that eliminates double-conversion and a few others that you wouldn't have to waste your time with if you don't want to

**Comment by michaeleisel** [comment]
yeah that means you can avoid folly

**Comment by michaeleisel** [comment]
(and also because folly simply calls into double-conversion, it's just a wrapper)

**Comment by lemire** [comment]
>  i think i used homebrew for a bunch of them? 

I want the code to be as cross-platform as possible. I always try to make sure that everything works at least on macOS and Linux. And I do not want to have to call a package manager, obviously.

I have used double-conversion in the past, but I quickly left it alone for obvious reasons (it is slow).

**Comment by lemire** [comment]
Note: Ideally, I want to be able to run tests in a container of some kind.

**Comment by lemire** [comment]
> yeah that means you can avoid folly

That's great news because it is a mess. They require pre-requites that you need to manually install... Of course, you can get it through system-specific package managers, but I don't want that.

**Comment by lemire** [comment]
@michaeleisel Ok. So though my results are different (I have my own independent test but I will soon copy yours), they agree with you so far in this specific fact: boost::spirit is fast. So that's good. 

Regarding your other concerns, I should point out that integer parsing is *much* faster than float parsing. So even if we somehow get at the integer values faster than the competition, it can only help us so much.

Of course, we want everything to be fast, but the it is really the floating-point logic that is expensive.

**Comment by lemire** [comment]
@michaeleisel Is it your belief that boost spirit offers exact parsing?

I cannot verify this exactness. Suppose that I have the following string "9.905764316189753e-306", then spirit gives me a double which maps back to  9.90576431618975073e-306 whereas there is a double with value 9.90576431618975326e-306. The two differs by an ULP of two.

In other words, it looks to be no better than what is already present in simdjson. Or, at least, both are inexact (tolerating a small ULP error).

**Comment by lemire** [comment]
Update: It does not appear like ScanadvDouble supports exact parsing.

**Comment by lemire** [comment]
@michaeleisel 

Ok. So we must still demonstrate that the code is correct. You wrote "I see - I feel like I've shown mathematically that they're sound, but is there a different sort of proof you're interested in?". Did you write this out somewhere that I miss...? Sure, we discussed the math behind compute_float_double, but that's trivial. It is the rest of the code that I am concerned with.

So let us take compute_float_64.  I realize that it must be inspired by the moderate path from [Alexander Huszagh](https://www.reddit.com/r/rust/comments/a6j5j1/making_rust_float_parsing_fast_and_correct/)? [But it is not obviously the same algorithm, is it?](https://github.com/Alexhuszagh/rust-lexical/blob/dc6421d6990aaf71f54f9a24876f932a782c273c/lexical-core/src/atof/algorithm/correct.rs) It would helpful if you could let me know how it relates.
 
The difficult part is the rounding, I am sure you agree.

So we have this piece of code...

```
 if (unlikely((upper & 0x1FF) == 0x1FF)) {
    *success = false;
	return 0;
 }
```

So if we have 9 trailing ones, we bail out. Presumably, we do so because we are not going to be able to round properly. The bailing out part is fine, but what is important next is to prove that we have to be sure that we round correctly if we do not bail out. 

Then, once we have done the test for a bail out... we extract a 54-bit mantissa from the upper variable. Then we round up based on last bit, like so...

```
mantissa += mantissa & 1;
mantissa >>= 1;
```

Presumably, we are certain that ``mantissa += mantissa & 1`` will not overflow. That is, we are certain that it is mantissa is different from 0x3fffffffffffff. That's believable, and I am sure we can demonstrate it with a bit of work, but I want to check... that's something you have determined?

Then, algorithmically, what this sequence of two lines does is that it *rounds up* based on the last bit. And that implements the "round half to even" routine.

No doubt I can figure it out, but can you give me your version of the math behind this?

**Comment by michaeleisel** [comment]
I'm trying to find my very first bit of math, I'm 99% sure I mentioned it in the project. But anyways, it described the following. Start with a `uint64_t` mantissa, `i`, and a `uint64_t` power of 10, `f`. the power of 10 may not fit in 64 bits, in which case we just take the most significant 64 bits. we also store a power of 2 to accompany it, indicating how many bits were truncated. Now, let `F` be the perfectly accurate version of `f`, which thus has infinite precision, and `f` is the truncated version. We always round down when truncating, so `F = f + e`, where  `e` is our error term, and `0 < e < 1` in the scope of this arithmetic. `i * F` is our desired result. `i * F = i * (f + e)`, so the maximum error from this multiplication is `i * e`, and `0 < i * e < i`. So if `i * F + i` is equal to `i * F` (after truncating to 54 bits for our final mantissa) then we know we can use `i * F`

Then, this is what I found after that, and mentioned in the other Github thing. Although, this second proof is not necessary for what we're doing, since we use `strtod`, when `i` is not accurate.
> by the way, this can also handle when the other term, `i`, is not accurate. let `I` be perfectly accurate `i`, and `I = i + e1`. let `F` be perfectly accurate power of 10/5, and `F = f + e2`. due to rounding down, `0 < e1, e2 < 1`. `I * F = (i + e1) * (f + e2) = i * f + i * e2 + f * e1 + e1 * e2`. `i * e2 + f * e1 + e1 * e2 < i + f + 1`. so currently we just add `i`, in this case we would add `i + f + 1`

Indeed, it's similar to the moderate path algorithm. I forget why I decided not to use 80-bit doubles, but I remember deciding that it was not the way to go after some testing. By the way, I benchmarked that rust library as well, I can share that code if you want. It turned out to be significantly slower than this implementation as well.

**Comment by michaeleisel** [comment]
That's a little more higher level than some of what you're reasoning about, but that's what I meant when I said the math was worked out

**Comment by michaeleisel** [comment]
Also, I didn't mention the exponent component in that math, but it's not too hard to work out, relatively easy compared to the mantissa

**Comment by michaeleisel** [comment]
By the way, if we find a battle-tested C/C++ implementation (or can easily port one) that does this exact operation, I'd be in support of that. I definitely want to think more carefully about `compute_float_128`, which is just a 128-bit version of what I mentioned above, and how we might replace it

File: src/generic/numberparsing.h [hunk 1]
```diff
@@ -119,160 +119,578 @@ static inline bool is_made_of_eight_digits_fast(const char *chars) {
...
+// the assumption is that power_index is in the [0, 2*308] interval
+static really_inline double compute_float_64(uint64_t power_index, uint64_t i,
+                                             bool negative, bool *success) {
+  components c =
+      power_of_ten_components[power_index]; // safe because power_index <= 2*308
+  uint64_t factor_mantissa = c.mantissa;
+  int lz = leading_zeroes(i);
+  i <<= lz;
+  __uint128_t large_mantissa = (__uint128_t)i * factor_mantissa;
+  uint64_t upper = large_mantissa >> 64;
```
Comment (michaeleisel) [inline src/generic/numberparsing.h]: this shift of 64 is basically just saying, "give me the higher-order 64-bit chunk of this `__uint128_t`

Comment (michaeleisel) [reply src/generic/numberparsing.h]: it can be done in assembly, at least on my intel machine, without any shifts

File: src/generic/numberparsing.h [hunk 2]
```diff
@@ -119,160 +119,578 @@ static inline bool is_made_of_eight_digits_fast(const char *chars) {
...
+  if (unlikely((upper & 0x1FF) == 0x1FF)) {
```
Comment (michaeleisel) [inline src/generic/numberparsing.h]: we have a result in the higher order 64 bits, `upper`, and we want to determine if adding `i` to it (as per the math mentioned in the comments for this PR) will change the most significant 54 bits. since `i` is a `uint64_t` and we're looking at the most significant 64 bits, `i` < `1 << 64`, so if adding 1 to upper, which is equivalent to `1 << 64` in the original arithmetic, doesn't change the 54 bits, then we're ok. the only way adding 1 could change those bits is if we had a long enough string of trailing 1's at the end of `upper`

Comment (michaeleisel) [reply src/generic/numberparsing.h]: also, we know that the leading 1 of `upper` will be either in most significant place, or the second most significant, because we've shifted `i` to have its leading 1 in its MSB and `f` as per the caches always has its leading 1 in its MSB. so, `2 ** 63 <= i, f <= 2 ** 64 - 1`. `(2 ** 63) ** 2 <= i * f <= (2 ** 64 - 1) ** 2`, thus `2 ** 126 <= i * f <= 2 ** 128 - ... < 2 ** 128`. i.e., it has the leading bit either 126 or 127 bits ahead of the start of the number

File: src/generic/numberparsing.h [hunk 3]
```diff
@@ -119,160 +119,578 @@ static inline bool is_made_of_eight_digits_fast(const char *chars) {
...
+    *success = false;
+    return 0;
   }
-  long double i;
-  if (*p == '0') { // 0 cannot be followed by an integer
-    ++p;
-    i = 0;
+  uint64_t mantissa = 0;
+  if (upper & (1ULL << 63)) { // todo: this branch needs to be explained
```
Comment (michaeleisel) [inline src/generic/numberparsing.h]: we've passed the `return 0`, so we know that what we have will not be affected by the inaccuracy of `f`, and now we just need to do bookkeeping. is the MSB 1? if so, shift by 10 instead of 9

File: src/generic/numberparsing.h [hunk 4]
```diff
@@ -119,160 +119,578 @@ static inline bool is_made_of_eight_digits_fast(const char *chars) {
...
+    mantissa = upper >> 9;
+    lz++;
   }
-  if ('.' == *p) {
-    ++p;
-    int fractional_weight = 308;
-    if (is_integer(*p)) {
-      unsigned char digit = *p - '0';
-      ++p;
+  mantissa += mantissa & 1;
```
Comment (michaeleisel) [inline src/generic/numberparsing.h]: implementing round-to-even as you mentioned

File: src/generic/numberparsing.h [hunk 5]
```diff
@@ -119,160 +119,578 @@ static inline bool is_made_of_eight_digits_fast(const char *chars) {
...
+  mantissa >>= 1;
+  mantissa &= ~(1ULL << 52);
```
Comment (michaeleisel) [inline src/generic/numberparsing.h]: zero out the implicit one, we'll need to `|` that bit with the exponent

File: src/generic/numberparsing.h [hunk 6]
```diff
@@ -119,160 +119,578 @@ static inline bool is_made_of_eight_digits_fast(const char *chars) {
...
+  // lz = 64, c.exp = -1087, and we get that real_exponent ends up
+  // being -1.
+  // We need to prove that we are always in range!!!
+  //////
+  uint64_t real_exponent =
+      c.exp + 1023 + (127 - lz); // todo: the RHS can be negative
+  // todo: At this point we should be able to prove that real_exponent is larger
+  // than 0 and strictly smaller than 0x7FF.
+  mantissa |= real_exponent << 52;
+  mantissa |= (((uint64_t)negative) << 63);
```
Comment (michaeleisel) [inline src/generic/numberparsing.h]: flip the bit to indicate if it's negative or not

File: src/generic/numberparsing.h (hunk 2 above)
Comment (michaeleisel) [reply src/generic/numberparsing.h]: could we have an additional check to see if the MSB is 1 and then use `0x3FF` instead of `0x1FF`? yes, but i found that in all those cases, it didn't affect the result. perhaps there was a bug in my implementation, but also it may be a mathematical connection i'm not seeing

**Comment by michaeleisel** [comment]
so that's the math behind it, happy to explain more. also happy to throw it out if we can find a similarly fast implementation out there with production testing

**Comment by lemire** [comment]
@michaeleisel 

Thank you.

Note: you seem to think that I am being critical and that I want to throw away this code. This is not true!!! I am just trying to fulfill the requirement that the implementation be transparently correct.

**Comment by lemire** [comment]
@michaeleisel 

I went looking through everything we have and I fail to see a justification for the round-to-even computation. I am guessing that you are assuming that we are never exactly between two binary64 value. But I see no argument for why that is not possible. If it happens then I guess half the time we will be rounding wrong.

Similarly, when rounding, we might overflow. Again, we know to show that this cannot happen... In this code, why can't mantissa be equal to exactly 0x3fffffffffffff? If this were to happen, we would overflow and get zero... possibly?

```
  mantissa += mantissa & 1;
  mantissa >>= 1;
```

**Comment by lemire** [comment]
@michaeleisel 

The obvious "fix" is to add checks. But if you have arguments as to why these checks are unnecessary, I'd like to have them.

**Comment by michaeleisel** [comment]
no worries, i was just reminded how much work this is and so wanted us not to get too attached to this implementation :)

for the rounding, i think you're right. most values of `f` in the lookup table have been rounded down, meaning that there will be some one somewhere out there, even if not in the most significant 64 bits. however, the small positive ones are exact. maybe i'm missing something else that would help us here, but i'm not sure. so i guess we'd want to check that it's not all trailing zeroes + an exact value of `f`. i was thinking, in that case, that we'd also want to think about the error term affecting the result even if adding it doesn't spill into the higher order bits, but the only time this is relevant is when `f` is exact, so that check is irrelevant anyways.

1<<53 is fine to store as a float, because even though it appears that the mantissa is all zeroes, there's still always the implicit leading one.

**Comment by lemire** [comment]
@michaeleisel 

> so i guess we'd want to check that it's not all trailing zeroes + an exact value of f.

That's easily done and only needed when we would otherwise round up. So this will not happen often, if at all. It may be possible to prove that it never happens... but for now, I will add a check.

> 1<<53 is fine to store as a float, because even though it appears that the mantissa is all zeroes, there's still always the implicit leading one.

Are you sure about this?

If mantissa is 0x3fffffffffffffUL and you do 

```
  mantissa += mantissa & 1;
  mantissa >>= 1;
```

Then you get 0x20000000000000 or `1<<53`.

That's not ok. The implicit 1 bit is (1<<52). The value `1<<53` contains 54 bits. You can only have a mantissa of 53 bits in binary64, with one implicit leading 1 bit.

Look at your own code...

```
mantissa &= ~(1ULL << 52);
```

It agrees with my description... The implicit bit is not at `1<<53`.

So unless you have some argument, we need a check there as well.

And I already needed to add a check for real_exponent overflowing. So that's three additional checks.

I am going to have to go through compute_float_128, but I suspect that I will need to add three additional checks as well.

These checks slow things down a bit, but it not as bad as it sounds since we are already using over 850 instructions per float parsed... so adding 10 more instructions here and there won't make a big difference.

Maybe you understand better now what I meant by "proving correct"? That is, even if I can't come up with a proof that these checks are needed, I would feel uneasy not having them there because someone might come up with an unlucky case (even if it is very unlucky) and we'd be caught with our pants down...

More later on this topic.

**Comment by michaeleisel** [comment]
yeah, these things were all thought through by me, i just made a few mistakes in the reasoning. for the overflow, i think you're right. it's possible that for all the values in the lookup table can't produce such a number when multiplied with any `i`, but for now it'd be easier and safer to just have a check

**Comment by lemire** [comment]
@michaeleisel Thank you. I will carefully grind it through. I will do some work to see if we can avoid adding too many new checks. As you point out, it may be possible that some of them are never needed.

**Comment by lemire** [comment]
@michaeleisel After adding all the checks that I think are needed, I get the following performance numbers of what is essentially your "canada.txt" benchmark:

```
simdjson    251.51 cycles/value 849.40 instructions/value branch misses/value 0.16  MB/s 267331480.26 cycles/byte 13.78 
newsimdjson 266.84 cycles/value 892.10 instructions/value branch misses/value 0.18  MB/s 251974549.01 cycles/byte 14.62 
```

That is, the "perfect/exact" number parsing is going to be slower than what is currently in master by about 5%.

This is not the worst case, however. In a scenario that I cooked up, the exact parsing is  3x slower. And there is not much that we can do to alleviate that unless we know how to beat `strtod` in the worst cases.

Still, I have a decent hope that we can be only 5% slower *in practice*. That would be quite acceptable.

**Comment by michaeleisel** [comment]
what's this worst case? is it always falling back to the 128-bit case? i'm wondering if we can replace the 128-bit thing with some open-source implementation that might even be better, e.g. diyFp. after all, it's been enough work to prove the 64-bit case :P

**Comment by lemire** [comment]
@michaeleisel  Do you have reasons to believe that `compute_float_128` is useful? In my experiments, throwing it away actually improve matters. And, obviously, it reduces the amount of code one has to worry about.

**Comment by lemire** [comment]
@michaeleisel 


> what's this worst case? 

Picking floats at random.

**Comment by lemire** [comment]
So here is why the 128-bit approach is not too helpful...


canada.txt scenario:

I report the fraction of the calls that were (finally) parsed by corresponding function:

```
double (fast) : 8.5 %,  64-bit: 91.0 %, 128-bit: 0.5 %, full fallback: 0.0 %, extra (stdlib early):0.4 %
```

So the 128-bit approach only handles 0.5% of the calls that make it out to the main part of the code. We have 0.4% of the calls that go to an early bail out (too many characters or something).

In my adversarial (random) scenario, I get...

```
double (fast) : 7.2 %,  64-bit: 92.6 %, 128-bit: 0.2 %, full fallback: 0.0 %, extra (stdlib early):2.3 %
```

It is even worse for the 128-bit approach. It is very unlikely that we make it there.

Basically, the 64-bit function is pretty good. It catches most things. So that's pretty much all you need.

**Comment by lemire** [comment]
Ok. Examining the results more carefully, I find that compute_float_128 might help a bit. It is a small gain (less than 5%) but given that this is simdjson, we probably want to keep this gain.

**Comment by lemire** [comment]
Ok. With "relatively little" extra effort, I wrote a full 128-bit version instead of the 96-bit version. On x64, it should need the same number of instructions...  and it will be cleaner to port to Visual Studio since we are dealing with clean 64-bit words (that is, there is no need to use uint128_t). I don't think that there is direct benefit to going wider (from 96-bit lookups to 128-bit lookups) because if 96 bits are not enough, it is not likely that 128 bits will save the day. But the speed should be more or less the same. One benefit of the older approach is that it shared the same lookup table whereas I created a fresh new one (to be clean). 

Anyhow, with that in place, we almost never fall back to the standard library. There are very unlucky cases corresponding to the case where it is hard to determine whether we are right in between two floats... but they are so uncommon at this point... much less than one percent.

Other cases where we fall to the standard library regularly is when there are too many digits or when the exponent is crazy. The latter cases are often subnormals or overflows...  They are all cases where I don't care if parsing is slow.

**Comment by lemire** [comment]
My main todo now is to integrate the code I privately hacked into simdjson, and this implies handling portability issue and so forth.

Along this, as a secondary today, is to check that the new tests I have intergrated in the code are needed. Two of these new tests never, ever fire no matter how many floats I throw at the code. So maybe they are unneeded. They don't harm the performance measurably, but they make the code uglier. I'd want to remove them if I could.

I think I should be able to get this done next week.

**Comment by michaeleisel** [comment]
Nice. I'd love to take all the unit tests we can find from other open-source double parsing libraries as well (with appropriate licensing of course)

**Comment by lemire** [comment]
@michaeleisel Ah. We don't need compute_float_128 at all. We can do an additional multiplication within compute_float_64 when the precision is not good enough. This massively save on the number of lines of code (and it may be marginally more efficient). This means that just compute_float_64 by itself can cover 99.9% of the parsing in practice.

**Comment by lemire** [comment]
@michaeleisel The overflow that I worried about *can* happen, but only when the leading bit of the product is 1. So it is a scenario where we should not have shifted down so much earlier. Thankfully, we can recover because we know what the answer (mantissa) has to be: (1ULL << 52).

**Comment by lemire** [comment]
@michaeleisel Ok. So it definitively does happen that you can end up right between two floats. I was able to derive actual cases. It is stupidly improbable (it would be hard to catch with random tests), but we have to check.

I think I have an idea on how we can avoid becoming so much slower than the current simdjson in my adversarial scenario (random floats). More soon.

**Comment by michaeleisel** [comment]
What’s your recovery plan exactly for the overflow?

And that’s good to hear about the adversarial case

**Comment by lemire** [comment]
@michaeleisel 

> What’s your recovery plan exactly for the overflow?

My current guess is that the proper thing to do is to just set the mantissa to a leading 1 followed by zero, while decrementing the number of leading zeroes. But I'd like to discuss this scenario some more. I'll be running additional tests also.

**Comment by lemire** [comment]
@michaeleisel Ok. So it is good. You can pretty much avoid having to call strtod except if you have denormal numbers or you fall very near in-between two floats.

In the worst case, it can still be much more expensive than our existing (simdjson) routine *even though* we are almost never calling the standard library. But I don't know exactly what accounts for the extra cost. More benchmarking will be needed.

**Comment by michaeleisel** [comment]
yeah that recovery plan sounds reasonable, i suppose it's pretty easy to check for this overflow. as for when we have to fall back to strtod, we can actually figure out when we get all 0's whether or not we have to fall back. either the power of 10 "mantissa" is exact, i.e. for values around 10^1 to 10^18, in which case it's perfectly between the two floats. or it's an underestimate, since we always round down in that lookup table, in which case to round up. but i'm not sure if it's worth it to add checks for this

by the way, it could be worth it, if we aggregate a bunch of unit tests from other floating point libs, to package them as an open-source project. it can be a big pain point for people

**Comment by lemire** [comment]
@michaeleisel 

- [Fast float parsing in practice](https://lemire.me/blog/2020/03/10/fast-float-parsing-in-practice/) (blog post)
- [benchmark/library (github)](https://github.com/lemire/fast_double_parser)

**Comment by lemire** [comment]
Next step for me is to integrate into simdjson and see what happens.

@michaeleisel  Please see the blog post where I credit you.

**Comment by michaeleisel** [comment]
Looks good! :)

**Comment by lemire** [comment]
I am closing this PR in favour of https://github.com/lemire/simdjson/pull/558
